src/test/flatfile_tests.cpp

```diff
@@ -0,0 +1,123 @@
+}
+
+BOOST_AUTO_TEST_CASE(flatfile_open)
+{
+    auto data_dir = SetDataDir("flatfile_test");
+    FlatFileSeq seq(data_dir, "a", 16 * 1024);
+
+    std::string line1("A purely peer-to-peer version of electronic cash would allow online "
+                      "payments to be sent directly from one party to another without going "
+                      "through a financial institution.");
+    std::string line2("Digital signatures provide part of the solution, but the main benefits are "
+                      "lost if a trusted third party is still required to prevent double-spending.");
+
+    size_t pos1 = 0;
+    size_t pos2 = pos1 + GetSerializeSize(line1, CLIENT_VERSION);
+
+    // Write first line to file.
+    {
+        CAutoFile file(seq.Open(FlatFilePos(0, pos1)), SER_DISK, CLIENT_VERSION);
+        file << LIMITED_STRING(line1, 256);
+    }
+
+    // Attempt to append to file opened in read-only mode.
+    {
+        CAutoFile file(seq.Open(FlatFilePos(0, pos2), true), SER_DISK, CLIENT_VERSION);
+        BOOST_CHECK_THROW(file << LIMITED_STRING(line2, 256), std::ios_base::failure);
+    }
+
+    // Append second line to file.
+    {
+        CAutoFile file(seq.Open(FlatFilePos(0, pos2)), SER_DISK, CLIENT_VERSION);
+        file << LIMITED_STRING(line2, 256);
+    }
+
+    // Read text from file in read-only mode.
+    {
+        std::string text;
+        CAutoFile file(seq.Open(FlatFilePos(0, pos1), true), SER_DISK, CLIENT_VERSION);
+
+        file >> LIMITED_STRING(text, 256);
+        BOOST_CHECK_EQUAL(text, line1);
+
+        file >> LIMITED_STRING(text, 256);
+        BOOST_CHECK_EQUAL(text, line2);
+    }
+
+    // Read text from file with position offset.
+    {
+        std::string text;
+        CAutoFile file(seq.Open(FlatFilePos(0, pos2)), SER_DISK, CLIENT_VERSION);
+
+        file >> LIMITED_STRING(text, 256);
+        BOOST_CHECK_EQUAL(text, line2);
+    }
+
+    // Ensure another file in the sequence has no data.
+    {
+        std::string text;
+        CAutoFile file(seq.Open(FlatFilePos(1, pos2)), SER_DISK, CLIENT_VERSION);
+        BOOST_CHECK_THROW(file >> LIMITED_STRING(text, 256), std::ios_base::failure);
+    }
+}
+
+BOOST_AUTO_TEST_CASE(flatfile_allocate)
+{
+    auto data_dir = SetDataDir("flatfile_test");
+    FlatFileSeq seq(data_dir, "a", 100);
+
+    bool out_of_space;
+
+    BOOST_CHECK_EQUAL(seq.Allocate(FlatFilePos(0, 0), 1, out_of_space), 100);
+    BOOST_CHECK_EQUAL(fs::file_size(seq.FileName(FlatFilePos(0, 0))), 100);
+    BOOST_CHECK(!out_of_space);
+
+    BOOST_CHECK_EQUAL(seq.Allocate(FlatFilePos(0, 99), 1, out_of_space), 0);
+    BOOST_CHECK_EQUAL(fs::file_size(seq.FileName(FlatFilePos(0, 99))), 100);
+    BOOST_CHECK(!out_of_space);
+
+    BOOST_CHECK_EQUAL(seq.Allocate(FlatFilePos(0, 99), 2, out_of_space), 101);
+    BOOST_CHECK_EQUAL(fs::file_size(seq.FileName(FlatFilePos(0, 99))), 200);
+    BOOST_CHECK(!out_of_space);
```

Comment:
Good idea.

Reply:
I couldn't get this test passing on Travis... tried setting `too_big` to a petabyte as well, and there was still some issue.

Reply:
The reason is that `size_t` is 32 bit and there is more than 2^32 bytes available on the system, resulting in an overflow. The Allocate method should probably take a `uint64_t` to avoid this.